arch/risc-v/include/arch.h

```diff
@@ -56,6 +56,21 @@
 #  define ARCH_SPGTS          (ARCH_PGT_MAX_LEVELS - 1)
 #endif
 
+/* Can be used by assembly code to access the structure, example:
+ *
+ * Get percpu structure:
+ * 1: csrr    a0, CSR_SCRATCH
+ *
+ * Get hartid:
+ * 2: REGLOAD a0, RISCV_PERCPU_HARTID(a0)
```

Comment:
it isn't safe to do the action in two steps since the thread might move to the new cpu between them.

Reply:
It's OK for now, mainly used in exception_common:
https://github.com/apache/nuttx/blob/7036098d237383ae80c91f710cab33fb2299da55/arch/risc-v/src/common/riscv_exception_common.S#L174-L182

Reply:
But the location of percpu remains same, so reading TCB via CSR_SCRATCH->tcb will remain atomic?

Holding the TCB is not safe though, unless we are in critical section...

Reply:
Yes, you are right. We need to use TP for this, but I don't think it is possible for FLAT mode.